### PR TITLE
feat(pairing): add pairing screens and functionality 

### LINKED ISSUE #6617: Integrate with pairing lib for phase 1
The idea is to land the phase 1 designs ASAP so we can begin testing against the prod servers on real devices. We will initially have little automated testing since we know the flow is going to change.

ref https://github.com/mozilla/fxa/issues/317
blocked by #6668

### PR DESCRIPTION
**Pull request by rfk**
This is https://github.com/mozilla/fxa-content-server/pull/6404 re-opened as a `feature.*` branch for test deployment on fxa-dev.

fixes #6617

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: adbeb17b42acc6121bce587cd453d726ff5c9251
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 28ea920244c68d8f37e18b2f30158b9fff741fd5
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 137d55170259bd34c04e080aa57cdf26b61e10dd
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 6441bca2d18987173eb2c1b98a4c30fb3b58b4d6
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

File: app/scripts/models/reliers/supplicant.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  keys_jwk: Vat.keysJwk().required().renameTo('keysJwk'),
+  redirect_uri: Vat.url().required().renameTo('redirectUri'),
+  response_type: Vat.string().required().valid('token', 'code').renameTo('responseType'),
+  scope: Vat.string().required().min(1),
+  state: Vat.string().required()
+};
+
+const SUPPLICANT_HASH_PARAMETER_SCHEMA = {
+  channel_id: Vat.hex().len(32).required().renameTo('channelId'),
+  channel_key: Vat.string().len(32).required().renameTo('channelKey')
```
Comment (vladikoff) [inline app/scripts/models/reliers/supplicant.js]: this should be 43, from the HKDF result

File: app/scripts/models/reliers/supplicant.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+};
+
+var SUPPLICANT_QUERY_PARAM_SCHEMA = {
+  access_type: Vat.accessType().renameTo('accessType'),
+  client_id: Vat.clientId().required().renameTo('clientId'),
+  code_challenge: Vat.codeChallenge().required().renameTo('codeChallenge'),
+  code_challenge_method: Vat.codeChallengeMethod().required().renameTo('codeChallengeMethod'),
+  keys_jwk: Vat.keysJwk().required().renameTo('keysJwk'),
+  redirect_uri: Vat.url().required().renameTo('redirectUri'),
+  response_type: Vat.string().required().valid('token', 'code').renameTo('responseType'),
```
Comment (vladikoff) [inline app/scripts/models/reliers/supplicant.js]: @rfk should we remove it for the pairing here?

Comment (rfk) [reply app/scripts/models/reliers/supplicant.js]: Sorry, I'm not sure what "it" refers to here.  If it's the `redirect_uri` field, then yes, that seems reasonable to me on the Authority side.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 9b23be2355392c37830740c1597cfcfbb54a06a2
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 9b23be2355392c37830740c1597cfcfbb54a06a2
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 50f816cb0ddd2789d826cdf23633326178de22e4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

File: app/scripts/models/reliers/supplicant.js (hunk 2 above)
Comment (vladikoff) [reply app/scripts/models/reliers/supplicant.js]: Sorry, `response_type`

Comment (vladikoff) [reply app/scripts/models/reliers/supplicant.js]: from mtg: `redirect_uri` can also probably go!

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 9b23be2355392c37830740c1597cfcfbb54a06a2
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 50f816cb0ddd2789d826cdf23633326178de22e4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

File: app/scripts/models/auth_brokers/pairing/authority.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  getOAuthResult(account, options = {}) {
+    return Promise.resolve().then(() => {
+      const keysJwe = this.get('keysJwe');
+      if (keysJwe) {
+        return keysJwe;
+      }
+
+      return this.request(this._notificationChannel.COMMANDS.PAIR_REQUEST_KEYS_JWE);
+    }).then(keysJwe => {
+      this.set({ keysJwe });
```
Comment (rfk) [inline app/scripts/models/auth_brokers/pairing/authority.js]: The desktop code was sending an object like `{ keys_jwe: JWE_VALUE }` but this code is expecting just `JWE_VALUE`; I've managed to get it working by changing desktop code to send just `JWE_VALUE` but TBH I think expecting an object feels cleaner...

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 82da8c73c2605f7ac2a67c7e54656b35b726ab56
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 82da8c73c2605f7ac2a67c7e54656b35b726ab56
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: e299dcdddbeb8fe0b3b93ab24707b16a182174d1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 82da8c73c2605f7ac2a67c7e54656b35b726ab56
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 903c87ce99b4012b29f38d13442587ea5840e5c4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 82da8c73c2605f7ac2a67c7e54656b35b726ab56
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 903c87ce99b4012b29f38d13442587ea5840e5c4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

File: app/scripts/lib/app-start.js [hunk 4]
```diff
@@ -259,7 +263,23 @@ Start.prototype = {
       // as the OAuth check - when users sign up for Sync
       // and verify in a 2nd browser, all we have to know the user
       // is completing a Sync flow is `service=sync`.
-      if (this._isOAuth()) {
+
+      if (this.isDevicePairingAsAuthority()) {
+        relier = new SupplicantRelier({}, {
```
Comment (shane-tomlinson) [inline app/scripts/lib/app-start.js]: Split the responsibilities of the AuthorityRelier from the SupplicantRelier.

File: app/scripts/lib/channel-server-client.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  /**
+   * Create a new channel on the channel server and create
+   * a channel encryption key. `channelId` and `channelKey`
+   * will be set as an attribute of the model.
+   *
+   * Used by the `auth` device.
+   */
+  createNew () {
```
Comment (shane-tomlinson) [inline app/scripts/lib/channel-server-client.js]: This is no longer needed.

File: app/scripts/lib/channel-server-client.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+const CONFIRMATION_CODE_INFO = 'identity.mozilla.com/picl/v1/pair/confirmation-code';
+const CONFIRMATION_CODE_LENGTH = 4;
+
+export default class ChannelServerClient extends Model {
+  constructor (attrs = {}, options = {}) {
+    if (attrs.channelId) {
+      attrs.type = 'supp';
+    } else {
+      attrs.type = 'auth';
```
Comment (shane-tomlinson) [inline app/scripts/lib/channel-server-client.js]: Remove the auth portions of the client, it's no longer needed since Desktop does all that work.

File: app/scripts/lib/channel-server-client.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+      .then(keysJwk => a256gcm.decrypt(ciphertext, keysJwk))
+      .then(result => JSON.parse(result));
+  }
+
+  _encrypt (envelope) {
+    return this._getChannelJwk()
+      .then(keysJwk => a256gcm.encrypt(JSON.stringify(envelope), keysJwk));
+  }
+
+  _generateChannelKey () {
```
Comment (shane-tomlinson) [inline app/scripts/lib/channel-server-client.js]: No longer needed, Desktop provides this info.

File: app/scripts/lib/channel-server-client.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    const channelKeyBuffer = Buffer.from(base64.base64UrlToBase64(channelKey), 'base64');
+    const channelIdBuffer = Buffer.from(channelId, 'utf8');
+    const infoBuffer = Buffer.from('identity.mozilla.com/picl/v1/pair/encryption-key', 'utf8');
+
+    console.log('channelKeyBuffer', channelKeyBuffer);
+    console.log('channelIdBuffer', channelIdBuffer);
+    console.log('infoBuffer', infoBuffer);
+    return Promise.all([
+      hkdf(channelKeyBuffer, channelIdBuffer, 'identity.mozilla.com/picl/v1/pair/encryption-key', DEVICE_PAIRING_CHANNEL_KEY_BYTES),
+      /*hkdf(channelKeyBuffer, channelIdBuffer, CONFIRMATION_CODE_INFO, CONFIRMATION_CODE_LENGTH),*/
```
Comment (shane-tomlinson) [inline app/scripts/lib/channel-server-client.js]: Put the confirmation code hkdf back in.

File: app/scripts/lib/crypto/ecdh.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import importFxaCryptoDeriver from './deriver';
```
Comment (shane-tomlinson) [inline app/scripts/lib/crypto/ecdh.js]: This might no longer be needed since Desktop does this work now.

File: app/scripts/lib/crypto/scoped-keys.js [hunk 10]
```diff
@@ -6,20 +6,18 @@
  * Derive scope-specific keys from account master keys.
  */
 
-
-function importFxaCryptoDeriver() {
-  return import(/* webpackChunkName: "fxaCryptoDeriver" */ 'fxaCryptoDeriver');
-}
+import ecdh from './ecdh';
```
Comment (shane-tomlinson) [inline app/scripts/lib/crypto/scoped-keys.js]: Can probably revert this change since Desktop takes care of creating the final keys_jwe and ecdh is no longer used.

File: app/scripts/lib/router.js [hunk 11]
```diff
@@ -117,6 +117,13 @@ const Router = Backbone.Router.extend({
     'oauth/force_auth(/)': createViewHandler(ForceAuthView),
     'oauth/signin(/)': 'onSignIn',
     'oauth/signup(/)': 'onSignUp',
+    'pair/auth(/)': createViewHandler('pairing/pair_auth'),
```
Comment (shane-tomlinson) [inline app/scripts/lib/router.js]: This route is no longer needed.

File: app/scripts/lib/router.js [hunk 12]
```diff
@@ -117,6 +117,13 @@ const Router = Backbone.Router.extend({
...
+    'pair/auth/allow(/)': createViewHandler('pairing/pair_auth_allow'),
+    'pair/auth/complete(/)': createViewHandler('pairing/pair_auth_complete'),
+    'pair/supp(/)': createViewHandler('pairing/pair_supp'),
+    'pair/supp/allow(/)': createViewHandler('pairing/pair_supp_allow'),
+    'pair/supp/complete(/)': createViewHandler('pairing/pair_supp_complete'),
```
Comment (shane-tomlinson) [inline app/scripts/lib/router.js]: I don't *think* this route is ever used, the OAuth flow redirects back to the app before showing this.

File: app/scripts/models/account.js [hunk 13]
```diff
@@ -1069,16 +1069,32 @@ const Account = Backbone.Model.extend({
...
+     * @param {String} [password]
+     * @param {Object} [relier] only needed if password is passed
      * @returns {Promise} that resolves with the account keys, if they
      *   can be generated, resolves with null otherwise.
      */
-  accountKeys () {
-    if (! this.canFetchKeys()) {
-      return Promise.resolve(null);
-    }
+  accountKeys (password, relier) {
```
Comment (shane-tomlinson) [inline app/scripts/models/account.js]: This change can be reverted since Desktop does the work of getting the keys.

File: app/scripts/models/pairing/auth-state-machine.js [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+    this.navigate('pair/auth/complete');
+  }
+}
+
+class AuthStateMachine extends PairingFlowStateMachine {
+  constructor(attrs, options = {}) {
+    super(attrs, options);
+
+    if (options.useWebChannel) {
```
Comment (shane-tomlinson) [inline app/scripts/models/pairing/auth-state-machine.js]: This option is no longer needed, remove it from the options when creating the state machine too.

File: app/scripts/models/reliers/supplicant.js (hunk 2 above)
Comment (shane-tomlinson) [reply app/scripts/models/reliers/supplicant.js]: @vladikoff this schema is the query param schema when loading the supplicant, I think it should stay here so we can redirect back to the app. The `SENT_TO_AUTHORITY_FROM_SUPPLICANT_SCHEMA` no longer includes either `redirect_uri` or `response_type`.

File: app/scripts/models/reliers/supplicant.js [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    this.importSearchParamsUsingSchema(SUPPLICANT_QUERY_PARAM_SCHEMA, OAuthErrors);
+    return this._setupOAuthRPInfo();
+  }
+
+  wantsKeys () {
+    return true;
+  }
+
+  validateApprovalData (approvalData) {
+    const { code, redirect, state } = approvalData;
```
Comment (shane-tomlinson) [inline app/scripts/models/reliers/supplicant.js]: Only code and state are sent, no redirect.

File: app/scripts/views/pairing/pair_auth_allow.js [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+      email: this.getAccount().get('email')
+    });
+  }
+
+  submit () {
+    return this.invokeBrokerMethod('afterPairAuthAllow', this.getAccount());
+  }
+
+  cancel () {
+    alert('cancel flow');
```
Comment (shane-tomlinson) [inline app/scripts/views/pairing/pair_auth_allow.js]: Send a WebChannel message when this occurs.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 82da8c73c2605f7ac2a67c7e54656b35b726ab56
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 903c87ce99b4012b29f38d13442587ea5840e5c4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 82da8c73c2605f7ac2a67c7e54656b35b726ab56
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 903c87ce99b4012b29f38d13442587ea5840e5c4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 82da8c73c2605f7ac2a67c7e54656b35b726ab56
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 903c87ce99b4012b29f38d13442587ea5840e5c4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 82da8c73c2605f7ac2a67c7e54656b35b726ab56
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 903c87ce99b4012b29f38d13442587ea5840e5c4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by eoger** [comment]
How do you feel about adding a version identifier in the QR code created by the auth and/or in the first `pair:supp:request` message sent by the supp.
